### ISSUE TITLE
Counterintuitive resizability behaviors [on x11]

### ISSUE THREAD

**Issue by 64kramsystem** [rust-windowing/winit#2306]
I'm using `winit` (0.26.x) as underlying library of a game, which I'm developing on Linux/x11. On startup, I need to resize the window, and prevent the user from resizing it.

Now, there are two problems:

1. `Window#set_resizable(false)` prevents also programmatic changes (in other words, it will cause subsequent calls to `set_inner_size()` won't have any effect); this is unexpected - I'd expect only user resizing to be prevented, not also programmatic one.

2. Invoking `set_inner_size()` and immediately after `set_resizable(false)`, will cause resizing to be disabled _without_ the window size to be changed; in order to change the size, a pause needs to be taken between the two statements (e.g. stepping through in a debugging session); I actually worked this around by putting `set_resizable()` in a resizing event callback.

It's not clear to me if the behaviors above are intended or they're bugs. If they're intended, I think they should be documented, as they're not obvious (I'd even classify them as counterintuitive).

**Comment by rib** [comment]
I agree that intuitively I'd only expect `set_resizable(false)` to request that the window manager should block the user from resizing the window, and I think that's the intent of the winit API.

The X11 backend does try to specifically account for `set_inner_size()` being called after `set_resizable(false)`, so I don't think the behaviour you're currently seeing is intentional.

This looks like a race condition due to `set_inner_size()` not caching the set size locally as a "pending" inner size that can be referenced by `set_resizable()`. On X11 the winit window's inner size only changes asynchronously, depending on whether the window manager decides to honor the request that's sent when `set_inner_size()` is called. Ideally though `set_resizable()` should be able to see that there is a "pending" (or something similar) inner size that has been requested and when it tries to disable resizing by setting the min/max to be the same in WM_NORMAL_HINTS it should request the pending size, not the
last size that was notified by the server.

For example the order of events can currently look like this:

| timeline of events/actions within winit           | winit's inner_size |
|---------------------------|---------------------|
| \<initial state\>                                               | 200x200|
| set_inner_size(300x300)                           | 200x200|
|  --- Send Resize Window request (300x300)    (i.e. just send async request)            | 200x200       |
|set_resizable(false)                                  | 200x200|
| --- read inner_size = 200x200               | 200x200|
| --- set WM_NORMAL_HINTS min/max = 200x200 | 200x200|
| handle configure notify from x server: 300x300    (if windows manager accepts original resize request before receiving new WM_NORMAL_HINTS)      | **300x300**|
| handle configure notify  from x server: 200x200    (if windows manager accepts new min/max size from WM_NORMAL_HINTS)      | **200x200**|

Incidentally, while investigating this, another small thing I noticed was that the implementation of `set_inner_size()` calls an `update_normal_hints` utility in the case where you have previously called `set_resizable(false)` to send new min/max values for `WM_NORMAL_HINTS` to the window manager. This utility is implemented like this:
```rust
fn update_normal_hints<F>(&self, callback: F) -> Result<(), XError>
    where
        F: FnOnce(&mut util::NormalHints<'_>),
    {
        let mut normal_hints = self.xconn.get_normal_hints(self.xwindow)?;
        callback(&mut normal_hints);
        self.xconn
            .set_normal_hints(self.xwindow, normal_hints)
            .flush()
    }
```
Which makes a synchronous get request to the X server, to ask for the old values, followed by an asynchronous set. This X server round trip is really far from ideal since it blocks that thread (probably the event loop) until the x server can respond to the request and in this case is redundant since there's no need to know the old values if they are simply being overwritten. As far as I could see then all current usage of this `update_normal_hints` in the X11 backend doesn't need to query back the old values, so the `.get_normal_hints()` request can probably just be dropped from that code I think.

**Comment by 64kramsystem** [comment]
Ah, hello again :smile:  Thanks for the in-depth analysis! :star_struck:

**Comment by rib** [comment]
Heya :)

I started looking at cooking up a patch but investigating the current x11 issues for winit I'm a little concerned that there are some really big PRs that are outstanding that will effectively conflict with almost any change in the backend but at the same time it's not really clear how they are progressing.

One ofthe difficulties is that there's no official 'maintainer' for the x11 backend which means theres no clear path to get changes reviewed and upstreamed.

One of the PRs is close to 150 patches, including changes dating back two years from the look of it which is a bit concerning.

**Comment by kchibisov** [comment]
>Window#set_resizable(false) prevents also programmatic changes (in other words, it will cause subsequent calls to set_inner_size() won't have any effect); this is unexpected - I'd expect only user resizing to be prevented, not also programmatic one.

This is fixed on the lastest master. Try using it instead.

**Comment by rib** [comment]
Ah, interesting that the changes to account for calling `set_inner_size()` for a window marked as !resizable were quite a recent change to improve consistency.

For reference my investigation above was just based on an inspection of the backend in master, and I might be mistaken from not having tested yet it but my current expectation/interpretation is that it will still have this issue on master...

The issue I see is that `set_inner_size_physical()` only sends the new size directly to the X server by calling `XResizeWindow` which will happen asynchronously. So internally the 'inner_size' doesn't immediately change after calling `set_inner_size`. That then means if an application calls `set_resizable(false)` immediately after `set_inner_size()` then the implementation of `set_resizable(false)` is going to refer to the old/out-of-date inner_size and that old size will then be used to configure the min/max normal_hints.

Technically the size might briefly be configured to the new inner_size but it's also likely to be immediately reverted back to the old size so it it would look like the set_inner_size didn't do anything.

**Comment by 64kramsystem** [comment]
Hi there!

>> Window#set_resizable(false) prevents also programmatic changes (in other words, it will cause subsequent calls to set_inner_size() won't have any effect); this is unexpected - I'd expect only user resizing to be prevented, not also programmatic one.
> This is fixed on the lastest master. Try using it instead.

Fixed indeed! Thanks :smile: 

I've done a few tests, and these are the results:

1. `set_resizable` in main, and `set_inner_size` in window event hook: window now resizes
2. `set_resizable` in main, and `set_inner_size` immediately after (back to back): window now resizes
3.  `set_inner_size`in main, and `set_resizable` immediately after (back to back): window still does not resize

I don't know if 3. is expected (or even 2.!), but even the improvement 2. makes the code nicer now :smile:
